File: shortcut_analyzer.rb
#!/usr/bin/env ruby
# shortcut_analyzer.rb
# encoding: UTF-8

require 'json'
require 'csv'
require 'set'

COLORS = {
  reset: "\e[0m",
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

class Analyzer
  attr_reader :shortcuts, :apps, :categories, :conflicts

  def initialize(data)
    @shortcuts = data['shortcuts'] || []
    @apps = Hash.new { |h, k| h[k] = [] }
    @categories = Hash.new { |h, k| h[k] = [] }
    @conflicts = Hash.new { |h, k| h[k] = [] }
    index
  end

  def index
    @shortcuts.each do |s|
      app = s['app'] || 'Без приложения'
      @apps[app] << s
      cat = s['category'] || 'Без категории'
      @categories[cat] << s
      @conflicts[s['shortcut']] << s if s['shortcut'] && !s['shortcut'].empty?
    end
  end

  def search(query)
    q = query.downcase
    @shortcuts.select do |s|
      s['name'].downcase.include?(q) || s['shortcut'].downcase.include?(q)
    end
  end

  def get_conflicts
    @conflicts.select { |_, v| v.size > 1 }
  end

  def group_by_app
    @apps
  end

  def export_csv(filename)
    CSV.open(filename, 'w') do |csv|
      csv << ['name', 'shortcut', 'app', 'category']
      @shortcuts.each { |s| csv << [s['name'], s['shortcut'], s['app'], s['category']] }
    end
    puts colorize("Экспортировано в #{filename}", :green)
  end

  def export_json(filename)
    File.write(filename, JSON.pretty_generate({ 'shortcuts' => @shortcuts }))
    puts colorize("Экспортировано в #{filename}", :green)
  end

  def display(items = nil, group: false, conflicts_only: false, verbose: false)
    data = items || @shortcuts
    return puts colorize('Нет данных для отображения.', :yellow) if data.empty?
    if conflicts_only
      conf = get_conflicts
      if conf.empty?
        puts colorize('Конфликтов не найдено.', :green)
        return
      end
      puts colorize('🔍 Конфликты (одинаковые комбинации):', :bold)
      conf.each do |combo, list|
        puts colorize("  #{combo}:", :red)
        list.each { |s| puts "    - #{s['name']} (#{s['app']})" }
      end
      return
    end
    if group
      group_by_app.each do |app, items|
        puts colorize("\n📁 #{app} (#{items.size}):", :blue)
        items.each { |s| print_item(s, verbose) }
      end
    else
      data.each { |s| print_item(s, verbose) }
    end
  end

  def print_item(s, verbose)
    if verbose
      puts "  #{colorize(s['name'], :bold)} → #{colorize(s['shortcut'], :cyan)}  (приложение: #{s['app']}, категория: #{s['category']})"
    else
      puts "  #{s['name']} → #{s['shortcut']}"
    end
  end
end

def main
  file = nil
  search = nil
  export_file = nil
  group = false
  conflicts_only = false
  verbose = false

  i = 0
  while i < ARGV.size
    arg = ARGV[i]
    case arg
    when '-s' then search = ARGV[i+1]; i += 1
    when '-g' then group = true
    when '-c' then conflicts_only = true
    when '-e' then export_file = ARGV[i+1]; i += 1
    when '-v' then verbose = true
    when '-h', '--help'
      puts "Usage: ruby shortcut_analyzer.rb <file> [-s search] [-g] [-c] [-e file] [-v]"
      return
    else file = arg if file.nil?
    end
    i += 1
  end

  unless file
    puts colorize('Укажите файл конфигурации.', :red)
    exit 1
  end

  begin
    data = JSON.parse(File.read(file))
  rescue => e
    puts colorize("Ошибка загрузки файла: #{e.message}", :red)
    exit 1
  end

  analyzer = Analyzer.new(data)

  if search
    results = analyzer.search(search)
    if results.any?
      puts colorize("🔍 Найдено #{results.size} совпадений:", :bold)
      analyzer.display(results, group: group, verbose: verbose)
    else
      puts colorize('Ничего не найдено.', :yellow)
    end
    return
  end

  if conflicts_only
    analyzer.display(conflicts_only: true)
    return
  end

  if export_file
    if export_file.end_with?('.csv')
      analyzer.export_csv(export_file)
    elsif export_file.end_with?('.json')
      analyzer.export_json(export_file)
    else
      puts colorize('Неизвестный формат экспорта. Используйте .csv или .json', :red)
    end
    return
  end

  analyzer.display(group: group, verbose: verbose)
end

main if __FILE__ == $0

File: shortcut_analyzer.js
// shortcut_analyzer.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

class Analyzer {
    constructor(data) {
        this.shortcuts = data.shortcuts || [];
        this.apps = {};
        this.categories = {};
        this.conflicts = {};
        this.index();
    }

    index() {
        for (const s of this.shortcuts) {
            const app = s.app || 'Без приложения';
            if (!this.apps[app]) this.apps[app] = [];
            this.apps[app].push(s);

            const cat = s.category || 'Без категории';
            if (!this.categories[cat]) this.categories[cat] = [];
            this.categories[cat].push(s);

            if (s.shortcut) {
                if (!this.conflicts[s.shortcut]) this.conflicts[s.shortcut] = [];
                this.conflicts[s.shortcut].push(s);
            }
        }
    }

    search(query) {
        const q = query.toLowerCase();
        return this.shortcuts.filter(s =>
            s.name.toLowerCase().includes(q) ||
            s.shortcut.toLowerCase().includes(q)
        );
    }

    getConflicts() {
        const result = {};
        for (const [key, val] of Object.entries(this.conflicts)) {
            if (val.length > 1) result[key] = val;
        }
        return result;
    }

    groupByApp() {
        return this.apps;
    }

    exportCSV(filename) {
        const lines = ['name,shortcut,app,category'];
        for (const s of this.shortcuts) {
            lines.push(`${s.name},${s.shortcut},${s.app||''},${s.category||''}`);
        }
        fs.writeFileSync(filename, lines.join('\n'));
        console.log(colorize(`Экспортировано в ${filename}`, 'green'));
    }

    exportJSON(filename) {
        fs.writeFileSync(filename, JSON.stringify({ shortcuts: this.shortcuts }, null, 2));
        console.log(colorize(`Экспортировано в ${filename}`, 'green'));
    }

    display(items, group, conflictsOnly, verbose) {
        const data = items || this.shortcuts;
        if (data.length === 0) {
            console.log(colorize('Нет данных для отображения.', 'yellow'));
            return;
        }
        if (conflictsOnly) {
            const conf = this.getConflicts();
            const keys = Object.keys(conf);
            if (keys.length === 0) {
                console.log(colorize('Конфликтов не найдено.', 'green'));
                return;
            }
            console.log(colorize('🔍 Конфликты (одинаковые комбинации):', 'bold'));
            for (const combo of keys) {
                console.log(colorize(`  ${combo}:`, 'red'));
                for (const item of conf[combo]) {
                    console.log(`    - ${item.name} (${item.app || 'Без приложения'})`);
                }
            }
            return;
        }
        if (group) {
            const grouped = this.groupByApp();
            for (const [app, items] of Object.entries(grouped)) {
                console.log(colorize(`\n📁 ${app} (${items.length}):`, 'blue'));
                for (const item of items) {
                    this.printItem(item, verbose);
                }
            }
        } else {
            for (const item of data) {
                this.printItem(item, verbose);
            }
        }
    }

    printItem(item, verbose) {
        if (verbose) {
            console.log(`  ${colorize(item.name, 'bold')} → ${colorize(item.shortcut, 'cyan')}  (приложение: ${item.app || 'Без приложения'}, категория: ${item.category || 'Без категории'})`);
        } else {
            console.log(`  ${item.name} → ${item.shortcut}`);
        }
    }
}

function main() {
    const args = process.argv.slice(2);
    let file = null, search = null, exportFile = null;
    let group = false, conflictsOnly = false, verbose = false;

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === '-s' && i+1 < args.length) search = args[++i];
        else if (arg === '-g') group = true;
        else if (arg === '-c') conflictsOnly = true;
        else if (arg === '-e' && i+1 < args.length) exportFile = args[++i];
        else if (arg === '-v') verbose = true;
        else if (arg === '-h' || arg === '--help') {
            console.log('Usage: node shortcut_analyzer.js <file> [-s search] [-g] [-c] [-e file] [-v]');
            process.exit(0);
        } else if (!file) file = arg;
    }
    if (!file) {
        console.log(colorize('Укажите файл конфигурации.', 'red'));
        process.exit(1);
    }

    let data;
    try {
        data = JSON.parse(fs.readFileSync(file, 'utf8'));
    } catch (err) {
        console.log(colorize(`Ошибка загрузки файла: ${err.message}`, 'red'));
        process.exit(1);
    }

    const analyzer = new Analyzer(data);

    if (search) {
        const results = analyzer.search(search);
        if (results.length) {
            console.log(colorize(`🔍 Найдено ${results.length} совпадений:`, 'bold'));
            analyzer.display(results, group, false, verbose);
        } else {
            console.log(colorize('Ничего не найдено.', 'yellow'));
        }
        return;
    }

    if (conflictsOnly) {
        analyzer.display(null, false, true, false);
        return;
    }

    if (exportFile) {
        if (exportFile.endsWith('.csv')) {
            analyzer.exportCSV(exportFile);
        } else if (exportFile.endsWith('.json')) {
            analyzer.exportJSON(exportFile);
        } else {
            console.log(colorize('Неизвестный формат экспорта. Используйте .csv или .json', 'red'));
        }
        return;
    }

    analyzer.display(null, group, false, verbose);
}

main();
